src/transformers/modeling_prophetnet.py

```diff
@@ -0,0 +1,1461 @@
+        attn_probs = F.dropout(attn_weights, p=self.dropout, training=self.training,)
+
+        assert v is not None
+        attn_output = torch.bmm(attn_probs, v)
+        assert attn_output.size() == (bsz * self.num_heads, tgt_len, self.head_dim)
+        attn_output = attn_output.transpose(0, 1).contiguous().view(tgt_len, bsz, embed_dim)
+        attn_output = self.out_proj(attn_output)
+        if output_attentions:
+            attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len, src_len)
+        else:
+            attn_weights = None
+        return attn_output, attn_weights
+
+    def _use_saved_state(self, k, v, saved_state, key_padding_mask, static_kv, bsz):
+        # saved states are stored with shape (bsz, num_heads, seq_len, head_dim)
+        if "prev_key" in saved_state:
+            _prev_key = saved_state["prev_key"]
+            assert _prev_key is not None
+            prev_key = _prev_key.view(bsz * self.num_heads, -1, self.head_dim)
+            if static_kv:
+                k = prev_key
+            else:
+                assert k is not None
+                k = torch.cat([prev_key, k], dim=1)
+        if "prev_value" in saved_state:
+            _prev_value = saved_state["prev_value"]
+            assert _prev_value is not None
+            prev_value = _prev_value.view(bsz * self.num_heads, -1, self.head_dim)
+            if static_kv:
+                v = prev_value
+            else:
+                assert v is not None
+                v = torch.cat([prev_value, v], dim=1)
+        assert k is not None and v is not None
+        prev_key_padding_mask: Optional[Tensor] = saved_state.get("prev_key_padding_mask", None)
+        key_padding_mask = self._cat_prev_key_padding_mask(
+            key_padding_mask, prev_key_padding_mask, bsz, k.size(1), static_kv
+        )
+        return k, v, key_padding_mask
+
+    @staticmethod
+    def _cat_prev_key_padding_mask(
+        key_padding_mask: Optional[Tensor],
+        prev_key_padding_mask: Optional[Tensor],
+        batch_size: int,
+        src_len: int,
+        static_kv: bool,
+    ) -> Optional[Tensor]:
+        # saved key padding masks have shape (bsz, seq_len)
+        if prev_key_padding_mask is not None:
+            if static_kv:
+                new_key_padding_mask = prev_key_padding_mask
+            else:
+                new_key_padding_mask = torch.cat([prev_key_padding_mask, key_padding_mask], dim=1)
+
+        elif key_padding_mask is not None:
+            filler = torch.zeros(
+                batch_size,
+                src_len - key_padding_mask.size(1),
+                dtype=key_padding_mask.dtype,
+                device=key_padding_mask.device,
+            )
+            new_key_padding_mask = torch.cat([filler, key_padding_mask], dim=1)
+        else:
+            new_key_padding_mask = prev_key_padding_mask
+        return new_key_padding_mask
+
+
+class EncoderLayer(nn.Module):
```

Comment:
I think we have to split the class up into more classes. If you take a look at BERT: https://github.com/huggingface/transformers/blob/31da35cc8939e19a292cb7f15cad5c9a1ddf7f23/src/transformers/modeling_bert.py#L376 you can see that the `BertLayer` is much more modularized. 

Note that once we have the layer names defined, such as `self.fc1` we cannot really change them any more afterwards because the trained weights are tied to it. Have a more modularized layer has the advantage that it's much easier and cleaner to add new features change the logic at a later stage. 

Say we want to add another version which has the layer norm before the feed forward layers. As it is now we would have to add `if` statements which are not very clean. If we instead modularize the feed forward part directly, such as self.feed_forward = FeedForwardLayer(...)` we could later simply add another `FeedForwardLayerNew(...)` and do if config.... `self.feed_forward = FeedForwardLayer(..) else self.feed_forward = FeedForwardLayerNew(...)`. This allowed us to simply device many models from BERT and makes the model in general much more flexible. I strongly recommend to do this here as well...if this means that it does not fit with your current weights, we can simply write a conversion script to convert the weights (takes 15min, I can help you :-)) 
This has a big advantage if we decide to add new features to the model at a later stage...say for example the feed forward layers would change. As it is implemented now we could not